File: backend/src/routes/admin/assignments.py
from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.orm import Session
from typing import List, Optional
from datetime import datetime
from ...database import get_db
from ...models import Assignment, User, Challenge, AssignmentStatus
from ...schemas import AssignmentCreate, AssignmentUpdate, AssignmentResponse, AssignmentWithDetails, BulkAssignmentCreate
from ...utils.auth import get_current_admin_user

router = APIRouter(prefix="/api/admin/assignments", tags=["admin-assignments"])


@router.post("", response_model=AssignmentResponse, status_code=status.HTTP_201_CREATED)
def create_assignment(
    assignment_data: AssignmentCreate,
    db: Session = Depends(get_db),
    admin: User = Depends(get_current_admin_user)
):
    """Assign a challenge to a user (Admin only)"""
    # Verify user exists
    user = db.query(User).filter(User.id == assignment_data.user_id).first()
    if not user:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="User not found"
        )
    
    # Verify challenge exists
    challenge = db.query(Challenge).filter(Challenge.id == assignment_data.challenge_id).first()
    if not challenge:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Challenge not found"
        )
    
    # Check for existing active assignment
    existing = db.query(Assignment).filter(
        Assignment.user_id == assignment_data.user_id,
        Assignment.challenge_id == assignment_data.challenge_id,
        Assignment.status.in_([AssignmentStatus.ASSIGNED, AssignmentStatus.IN_PROGRESS])
    ).first()
    
    if existing:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="User already has an active assignment for this challenge"
        )
    
    assignment = Assignment(
        **assignment_data.dict(),
        assigned_by=admin.id
    )
    db.add(assignment)
    db.commit()
    db.refresh(assignment)
    return assignment


@router.post("/bulk", response_model=List[AssignmentResponse], status_code=status.HTTP_201_CREATED)
def bulk_create_assignments(
    bulk_data: BulkAssignmentCreate,
    db: Session = Depends(get_db),
    admin: User = Depends(get_current_admin_user)
):
    """Assign a challenge to multiple users at once (Admin only)"""
    # Verify challenge exists
    challenge = db.query(Challenge).filter(Challenge.id == bulk_data.challenge_id).first()
    if not challenge:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Challenge not found"
        )
    
    # Verify all users exist
    users = db.query(User).filter(User.id.in_(bulk_data.user_ids)).all()
    found_user_ids = {user.id for user in users}
    missing_user_ids = set(bulk_data.user_ids) - found_user_ids
    
    if missing_user_ids:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Users not found: {', '.join(map(str, missing_user_ids))}"
        )
    
    created_assignments = []
    skipped_users = []
    
    for user_id in bulk_data.user_ids:
        # Check for existing active assignment
        existing = db.query(Assignment).filter(
            Assignment.user_id == user_id,
            Assignment.challenge_id == bulk_data.challenge_id,
            Assignment.status.in_([AssignmentStatus.ASSIGNED, AssignmentStatus.IN_PROGRESS])
        ).first()
        
        if existing:
            skipped_users.append(user_id)
            continue
        
        assignment = Assignment(
            user_id=user_id,
            challenge_id=bulk_data.challenge_id,
            notes=bulk_data.notes,
            expires_at=bulk_data.expires_at,
            assigned_by=admin.id
        )
        db.add(assignment)
        created_assignments.append(assignment)
    
    if created_assignments:
        db.commit()
        for assignment in created_assignments:
            db.refresh(assignment)
    
    return created_assignments


@router.get("", response_model=List[AssignmentWithDetails])
def list_assignments(
    db: Session = Depends(get_db),
    admin: User = Depends(get_current_admin_user),
    user_id: Optional[int] = Query(None, description="Filter by user ID"),
    challenge_id: Optional[int] = Query(None, description="Filter by challenge ID"),
    status: Optional[str] = Query(None, description="Filter by status")
):
    """List all assignments with filters (Admin only)"""
    query = db.query(
        Assignment,
        User.email.label("user_email"),
        Challenge.name.label("challenge_name"),
        User.email.label("admin_email")
    ).join(
        User, Assignment.user_id == User.id
    ).join(
        Challenge, Assignment.challenge_id == Challenge.id
    ).join(
        User, Assignment.assigned_by == User.id, isouter=True
    )
    
    if user_id:
        query = query.filter(Assignment.user_id == user_id)
    if challenge_id:
        query = query.filter(Assignment.challenge_id == challenge_id)
    if status:
        query = query.filter(Assignment.status == status)
    
    results = query.all()
    
    # Format response
    assignments = []
    for assignment, user_email, challenge_name, admin_email in results:
        assignment_dict = {
            **assignment.__dict__,
            "user_email": user_email,
            "challenge_name": challenge_name,
            "admin_email": admin_email
        }
        assignments.append(assignment_dict)
    
    return assignments


@router.get("/{assignment_id}", response_model=AssignmentResponse)
def get_assignment(
    assignment_id: int,
    db: Session = Depends(get_db),
    admin: User = Depends(get_current_admin_user)
):
    """Get assignment details (Admin only)"""
    assignment = db.query(Assignment).filter(Assignment.id == assignment_id).first()
    if not assignment:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Assignment not found"
        )
    return assignment


@router.put("/{assignment_id}", response_model=AssignmentResponse)
def update_assignment(
    assignment_id: int,
    assignment_data: AssignmentUpdate,
    db: Session = Depends(get_db),
    admin: User = Depends(get_current_admin_user)
):
    """Update an assignment (Admin only)"""
    assignment = db.query(Assignment).filter(Assignment.id == assignment_id).first()
    if not assignment:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Assignment not found"
        )
    
    # Update only provided fields
    update_data = assignment_data.dict(exclude_unset=True)
    for field, value in update_data.items():
        setattr(assignment, field, value)
    
    # Update completed_at if status changed to completed
    if assignment_data.status == AssignmentStatus.COMPLETED.value and not assignment.completed_at:
        assignment.completed_at = datetime.utcnow()
    
    db.commit()
    db.refresh(assignment)
    return assignment


@router.delete("/{assignment_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_assignment(
    assignment_id: int,
    db: Session = Depends(get_db),
    admin: User = Depends(get_current_admin_user)
):
    """Delete an assignment (Admin only)"""
    assignment = db.query(Assignment).filter(Assignment.id == assignment_id).first()
    if not assignment:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Assignment not found"
        )
    
    db.delete(assignment)
    db.commit()
    return None

File: workers/src/routes/assignments.ts
import { Env, Assignment, User, Challenge } from '../types';
import { errorResponse, successResponse } from '../utils/response';
import { getUserIdFromRequest, requireAdmin } from '../utils/auth';

/**
 * Create a single assignment
 */
export async function createAssignment(request: Request, env: Env): Promise<Response> {
  try {
    const userId = await getUserIdFromRequest(request, env);
    if (!userId || !(await requireAdmin(userId, env))) {
      return errorResponse('Unauthorized - Admin access required', 403);
    }

    const body = await request.json();
    const { user_id, challenge_id, notes, expires_at } = body;

    // Verify user exists
    const user = await env.DB.prepare(
      'SELECT * FROM users WHERE id = ?'
    ).bind(user_id).first();

    if (!user) {
      return errorResponse('User not found', 404);
    }

    // Verify challenge exists
    const challenge = await env.DB.prepare(
      'SELECT * FROM challenges WHERE id = ?'
    ).bind(challenge_id).first();

    if (!challenge) {
      return errorResponse('Challenge not found', 404);
    }

    // Check for existing active assignment
    const existing = await env.DB.prepare(`
      SELECT * FROM assignments 
      WHERE user_id = ? AND challenge_id = ? 
      AND status IN ('assigned', 'in_progress')
    `).bind(user_id, challenge_id).first();

    if (existing) {
      return errorResponse('User already has an active assignment for this challenge', 400);
    }

    // Create assignment
    const result = await env.DB.prepare(`
      INSERT INTO assignments (
        user_id, challenge_id, assigned_by, notes, expires_at, assigned_at
      ) VALUES (?, ?, ?, ?, ?, datetime('now'))
      RETURNING *
    `).bind(
      user_id,
      challenge_id,
      userId,
      notes || null,
      expires_at || null
    ).first();

    return successResponse(result, 'Assignment created successfully');
  } catch (error) {
    console.error('Create assignment error:', error);
    return errorResponse('Failed to create assignment', 500);
  }
}

/**
 * Bulk create assignments
 */
export async function bulkCreateAssignments(request: Request, env: Env): Promise<Response> {
  try {
    const adminId = await getUserIdFromRequest(request, env);
    if (!adminId || !(await requireAdmin(adminId, env))) {
      return errorResponse('Unauthorized - Admin access required', 403);
    }

    const body = await request.json();
    const { user_ids, challenge_id, notes, expires_at } = body;

    if (!Array.isArray(user_ids) || user_ids.length === 0) {
      return errorResponse('user_ids must be a non-empty array', 400);
    }

    // Verify challenge exists
    const challenge = await env.DB.prepare(
      'SELECT * FROM challenges WHERE id = ?'
    ).bind(challenge_id).first();

    if (!challenge) {
      return errorResponse('Challenge not found', 404);
    }

    // Verify all users exist
    const placeholders = user_ids.map(() => '?').join(',');
    const { results: users } = await env.DB.prepare(
      `SELECT id FROM users WHERE id IN (${placeholders})`
    ).bind(...user_ids).all();

    const foundUserIds = new Set(users?.map((u: any) => u.id) || []);
    const missingUserIds = user_ids.filter(id => !foundUserIds.has(id));

    if (missingUserIds.length > 0) {
      return errorResponse(`Users not found: ${missingUserIds.join(', ')}`, 404);
    }

    // Create assignments for each user
    const created = [];
    const skipped = [];

    for (const userId of user_ids) {
      // Check for existing active assignment
      const existing = await env.DB.prepare(`
        SELECT * FROM assignments 
        WHERE user_id = ? AND challenge_id = ? 
        AND status IN ('assigned', 'in_progress')
      `).bind(userId, challenge_id).first();

      if (existing) {
        skipped.push(userId);
        continue;
      }

      // Create assignment
      const result = await env.DB.prepare(`
        INSERT INTO assignments (
          user_id, challenge_id, assigned_by, notes, expires_at, assigned_at
        ) VALUES (?, ?, ?, ?, ?, datetime('now'))
        RETURNING *
      `).bind(
        userId,
        challenge_id,
        adminId,
        notes || null,
        expires_at || null
      ).first();

      created.push(result);
    }

    return successResponse({
      created: created.length,
      skipped: skipped.length,
      skipped_user_ids: skipped,
      assignments: created
    }, `Created ${created.length} assignments, skipped ${skipped.length}`);
  } catch (error) {
    console.error('Bulk create assignments error:', error);
    return errorResponse('Failed to create assignments', 500);
  }
}

/**
 * Get all assignments (admin view)
 */
export async function getAllAssignments(request: Request, env: Env): Promise<Response> {
  try {
    const adminId = await getUserIdFromRequest(request, env);
    if (!adminId || !(await requireAdmin(adminId, env))) {
      return errorResponse('Unauthorized - Admin access required', 403);
    }

    const url = new URL(request.url);
    const userId = url.searchParams.get('user_id');
    const challengeId = url.searchParams.get('challenge_id');
    const status = url.searchParams.get('status');

    let query = `
      SELECT 
        a.*,
        u.email as user_email,
        c.name as challenge_name,
        admin.email as admin_email
      FROM assignments a
      LEFT JOIN users u ON a.user_id = u.id
      LEFT JOIN challenges c ON a.challenge_id = c.id
      LEFT JOIN users admin ON a.assigned_by = admin.id
      WHERE 1=1
    `;

    const params = [];

    if (userId) {
      query += ' AND a.user_id = ?';
      params.push(parseInt(userId));
    }

    if (challengeId) {
      query += ' AND a.challenge_id = ?';
      params.push(parseInt(challengeId));
    }

    if (status) {
      query += ' AND a.status = ?';
      params.push(status);
    }

    query += ' ORDER BY a.assigned_at DESC';

    // Log for debugging
    console.log(`Executing assignments query: ${query} with params: ${JSON.stringify(params)}`);

    const stmt = env.DB.prepare(query);
    const { results } = await (params.length > 0 ? stmt.bind(...params) : stmt).all();

    return successResponse(results || [], 'Assignments retrieved successfully');
  } catch (error: any) {
    console.error('Get assignments error:', error);
    return errorResponse(`Failed to get assignments: ${error.message || error.toString()}`, 500);
  }
}

/**
 * Get user's own assignments
 */
export async function getUserAssignments(request: Request, env: Env): Promise<Response> {
  try {
    const userId = await getUserIdFromRequest(request, env);
    if (!userId) {
      return errorResponse('Unauthorized', 401);
    }

    const { results } = await env.DB.prepare(`
      SELECT 
        a.*,
        c.name as challenge_name,
        c.description as challenge_description,
        c.difficulty as challenge_difficulty,
        c.duration_hours as challenge_duration_hours,
        (
          SELECT COUNT(*) FROM instances 
          WHERE assignment_id = a.id 
          AND status IN ('provisioning', 'running')
        ) as has_active_vm
      FROM assignments a
      JOIN challenges c ON a.challenge_id = c.id
      WHERE a.user_id = ?
      ORDER BY a.assigned_at DESC
    `).bind(userId).all();

    return successResponse(results || [], 'User assignments retrieved successfully');
  } catch (error) {
    console.error('Get user assignments error:', error);
    return errorResponse('Failed to get assignments', 500);
  }
}

/**
 * Update assignment
 */
export async function updateAssignment(request: Request, env: Env, assignmentId: string): Promise<Response> {
  try {
    const adminId = await getUserIdFromRequest(request, env);
    if (!adminId || !(await requireAdmin(adminId, env))) {
      return errorResponse('Unauthorized - Admin access required', 403);
    }

    const body = await request.json();
    const { status, notes, expires_at } = body;

    const assignment = await env.DB.prepare(
      'SELECT * FROM assignments WHERE id = ?'
    ).bind(assignmentId).first();

    if (!assignment) {
      return errorResponse('Assignment not found', 404);
    }

    // Build update query
    const updates = [];
    const params = [];

    if (status) {
      updates.push('status = ?');
      params.push(status);

      if (status === 'completed' && !(assignment as any).completed_at) {
        updates.push("completed_at = datetime('now')");
      }
    }

    if (notes !== undefined) {
      updates.push('notes = ?');
      params.push(notes);
    }

    if (expires_at !== undefined) {
      updates.push('expires_at = ?');
      params.push(expires_at);
    }

    if (updates.length === 0) {
      return errorResponse('No fields to update', 400);
    }

    params.push(assignmentId);

    const result = await env.DB.prepare(`
      UPDATE assignments 
      SET ${updates.join(', ')}
      WHERE id = ?
      RETURNING *
    `).bind(...params).first();

    return successResponse(result, 'Assignment updated successfully');
  } catch (error) {
    console.error('Update assignment error:', error);
    return errorResponse('Failed to update assignment', 500);
  }
}

/**
 * Delete assignment
 */
export async function deleteAssignment(request: Request, env: Env, assignmentId: string): Promise<Response> {
  try {
    const adminId = await getUserIdFromRequest(request, env);
    if (!adminId || !(await requireAdmin(adminId, env))) {
      return errorResponse('Unauthorized - Admin access required', 403);
    }

    const assignment = await env.DB.prepare(
      'SELECT * FROM assignments WHERE id = ?'
    ).bind(assignmentId).first();

    if (!assignment) {
      return errorResponse('Assignment not found', 404);
    }

    await env.DB.prepare('DELETE FROM assignments WHERE id = ?').bind(assignmentId).run();

    return successResponse(null, 'Assignment deleted successfully');
  } catch (error) {
    console.error('Delete assignment error:', error);
    return errorResponse('Failed to delete assignment', 500);
  }
}
